Remotion/Web/UnitTests/Core/Context/HttpContextStorageProviderWithAsyncLocalBasedFallbackTest.cs

```diff
@@ -0,0 +1,161 @@
+// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Collections;
+using System.Reflection;
+using System.Threading;
+using System.Web;
+using NUnit.Framework;
+using Remotion.Development.Web.UnitTesting.AspNetFramework;
+using Remotion.Web.Context;
+
+namespace Remotion.Web.UnitTests.Core.Context
+{
+  [TestFixture]
+  public class HttpContextStorageProviderWithAsyncLocalBasedFallbackTest
+  {
+    private HttpContextStorageProviderWithAsyncLocalBasedFallback _provider;
+    private HttpContext _testContext;
+
+    [SetUp]
+    public void SetUp ()
+    {
+      _testContext = HttpContextHelper.CreateHttpContext("x", "y", "z");
+      HttpContext.Current = _testContext;
+      _provider = new HttpContextStorageProviderWithAsyncLocalBasedFallback();
+    }
+
+    [Test]
+    public void DataRestoredAfterSafeContextBoundaryDisposed ()
+    {
+      _provider.SetData("test", 123);
+      using (_provider.OpenSafeContextBoundary())
+      {
+        var safeContextResult = _provider.GetData("test");
+        Assert.That(safeContextResult, Is.Null);
+      }
+
+      var result = _provider.GetData("test");
+      Assert.That(result, Is.EqualTo(123));
+    }
+
+    [Test]
+    public void SafeContextBoundaryDataUnavailableAfterDispose ()
+    {
+      using (_provider.OpenSafeContextBoundary())
+      {
+        _provider.SetData("test", 123);
+        var safeContextResult = _provider.GetData("test");
+        Assert.That(safeContextResult, Is.EqualTo(123));
+      }
+
+      var result = _provider.GetData("test");
+      Assert.That(result, Is.Null);
+    }
+
+    [Test]
+    public void SafeContextBoundaryCreationClearsData ()
+    {
+      _provider.SetData("test", 123);
+      _provider.OpenSafeContextBoundary();
```

Comment:
I think we should add a using here. It feels strange to leak the boundary during a test.